counsel.el

```diff
@@ -2063,6 +2063,65 @@ By default `counsel-bookmark' opens a dired buffer for directories."
    ("r" ,(counsel--apply-bookmark-fn #'counsel-find-file-as-root)
         "open as root")))
 
+;;** `counsel-bookmarked-directory'
+(defcustom counsel-bookmarked-directory-default-name "%s"
+  "Default bookmark name created by `counsel-bookmarked-directory-add."
+  :group 'ivy
+  :type 'string)
+
+(defun counsel-bookmarked-directory--path-separator ()
+  "Get the path separator for the current system."
+  (let* ((parent (expand-file-name "a"))
+         (i (length parent)))
+    (substring (expand-file-name "b" parent) i (1+ i))))
+
+(defconst counsel-bookmarked-directory-separator
+  (counsel-bookmarked-directory--path-separator))
+
+(defun counsel-bookmarked-directory--candidates ()
+  "Get a list of bookmarked directories sorted by file path."
+  (bookmark-maybe-load-default-file)
+  (cl-sort (cl-remove-if-not
+            (lambda (filename)
+              (string-suffix-p counsel-bookmarked-directory-separator filename))
+            (mapcar #'bookmark-get-filename bookmark-alist))
+           #'string<))
+
+;;;###autoload
+(defun counsel-bookmarked-directory-add (dir name)
+  "Add a bookmark to DIR with NAME."
+  (interactive (list default-directory
+                     (let ((path (abbreviate-file-name default-directory)))
+                       (read-from-minibuffer (format "Name of the new bookmark to %s: "
+                                                     path)
+                                             (format counsel-bookmarked-directory-default-name
+                                                     path)))))
```

Comment:
Why are `DIR` and `NAME` separate arguments, rather than a single file name argument?

Why are you using `read-from-minibuffer` instead of `read-directory-name`?

If `counsel-bookmarked-directory-default-name` is always going to be formatted based on `default-directory`, then what's its purpose? Manipulating file names as strings (e.g. by interpolation) is usually a bad idea, so why not provide a user option `counsel-bookmarked-directory-default` which will act as the second argument to `read-directory-name` instead?